server/utils/placements/index.ts

```diff
@@ -7,29 +7,40 @@ import paths from '../../paths/manage'
 import { hasPermission } from '../users/roles'
 
 export const actions = (placement: Cas1SpaceBooking, user: UserDetails) => {
-  const out = []
+  const actionList = []
+  const arrived = !!placement.actualArrivalDate
+  const departed = !!placement.actualDepartureDate
+  const nonArrival = !!placement.nonArrival
 
-  if (hasPermission(user, ['cas1_space_booking_record_keyworker'])) {
-    out.push({
+  if (!departed && !nonArrival && hasPermission(user, ['cas1_space_booking_record_keyworker'])) {
+    actionList.push({
       text: 'Assign keyworker',
       classes: 'govuk-button--secondary',
       href: paths.premises.placements.keyworker({ premisesId: placement.premises.id, placementId: placement.id }),
     })
   }
-  if (!placement.actualArrivalDate && hasPermission(user, ['cas1_space_booking_record_arrival'])) {
-    out.push({
+  if (!arrived && !nonArrival && hasPermission(user, ['cas1_space_booking_record_arrival'])) {
```

Comment:
I guess this and the next if/else conditional should both satisfy `!arrived && !nonArrival && !departed` so could be combined into one?

Reply:
Combining the tests probably makes the code less readable but OK.

Reply:
I think I was thinking of these conditionals as matching 'states' for the placement, effectively:

- Upcoming: no arrival or non-arrival recorded (show 'Record arrival' and 'Record non-arrival')
- Current: arrival recorded, no departure recorded (show 'Record departure')
- Historic: departure or non-arrival recorded (don't show anything?)

Similarly, there's a further conditional that checks `arrived && !nonArrival && !departed` -- surely if a placement has an arrival, it can't ever have a non-arrival as well?

Reply:
True. I was making the tests independent so that it would be easier to re-arrange the menu options in the future if required. You're quite right - you can only depart a placement that's arrived.